src/serac/numerics/functional/tensor.hpp

```diff
@@ -1185,43 +1200,187 @@ SERAC_HOST_DEVICE constexpr tensor<T, n> linear_solve(tensor<T, n, n> A, const t
     y        = tmp;
   };
 
-  tensor<double, n> x{};
+  auto U = A;
+  // initialize L to Identity
+  auto L = tensor<T, n, n>{};
+  // This handles the case if T is a dual number
+  // TODO: consider making a dense identity that is templated on type
+  // BT 05/09/2022
+  for (int i = 0; i < n; i++) {
+    if constexpr (is_dual_number<T>::value) {
+      L[i][i].value = 1.0;
+    } else {
+      L[i][i] = 1.0;
+    }
+  }
+  tensor<int, n> P(make_tensor<n>([](auto i) { return i; }));
 
   for (int i = 0; i < n; i++) {
     // Search for maximum in this column
-    double max_val = abs(A[i][i]);
+    double max_val = abs(get_value(U[i][i]));
 
     int max_row = i;
     for (int j = i + 1; j < n; j++) {
-      if (abs(A[j][i]) > max_val) {
-        max_val = abs(A[j][i]);
+      auto U_ji = get_value(U[j][i]);
+      if (abs(U_ji) > max_val) {
+        max_val = abs(U_ji);
         max_row = j;
       }
     }
 
-    swap(b[max_row], b[i]);
-    swap(A[max_row], A[i]);
+    swap(P[max_row], P[i]);
+    swap(U[max_row], U[i]);
+  }
 
-    // zero entries below in this column
+  for (int i = 0; i < n; i++) {
+    // zero entries below in this column in U
+    // and fill in L entries
     for (int j = i + 1; j < n; j++) {
-      double c = -A[j][i] / A[i][i];
-      A[j] += c * A[i];
-      b[j] += c * b[i];
-      A[j][i] = 0;
+      auto c  = U[j][i] / U[i][i];
+      L[j][i] = c;
+      U[j] -= c * U[i];
+      U[j][i] = T{};
     }
   }
 
-  // Solve equation Ax=b for an upper triangular matrix A
-  for (int i = n - 1; i >= 0; i--) {
-    x[i] = b[i] / A[i][i];
-    for (int j = i - 1; j >= 0; j--) {
-      b[j] -= A[j][i] * x[i];
+  return LuFactorization<T, n>{.P = P, .L = L, .U = U};
```

Comment:
just a C++ note: the aggregate (brace) initialization doesn't require the designators (i.e. `.P = ..., .L = ...`). We use that syntax in a couple places in serac because there are some arguments that are defaulted. So, this line could just be written 
```cpp
return LuFactorization<T, n>{P, L, U};
```
or (I think) even
```cpp
return {P, L, U};
```
since you declared the full return type in the function declaration.

Reply:
Indeed the second one works. Thanks for the suggestion. Incorportated in https://github.com/LLNL/serac/pull/691/commits/8cd4d0423dd683767e4547bee9b5fad4017c49ac.